### DISCUSSION TITLE
sliders with multiple windows

### DISCUSSION THREAD

**Discussion by antonfrancois** [marcomusy/vedo#527] [Q&A] [answered]
Hello

It is the follow-up of my previous : [question](https://github.com/marcomusy/vedo/discussions/514).
I am trying to make a slider that control two objects in respective subplots. I manage to make it work it when there is only one window `vedo.P lotter(N=1,...)` but when `N>=2` the slider does not show. 

I want the slider 'on top' of the two windows like in this sketch. How to do that ?
![IMG_20211110_151259](https://user-images.githubusercontent.com/6331304/141129521-27f6a76a-71fe-4f24-9f78-c351c9aec4d9.jpg)

Thank you again for the fasts answers and the great support you always give ! 
Anton

**Comment by marcomusy** [comment]
It's theoretically possible but you need to create a subwindow (renderer) just to host the slider.. otherwise you will have in one of the two. Check out the example:
`examples/basic/sliders2.py`

and for the creation of an overlapping renderer:
`examples/basic/multirenderers.py`
give it a try, If you don't manage I'll help you.

**Reply by antonfrancois** [reply]
I managed to make it works, and I don't want to add to much renderers. However there is a small issue 

 There where things that where confusing for me and I'll explain fully here for other newbies in my case.
If we want to have two windows, or renders we can set 
```python
 plt = vedo.Plotter(N=2,...)
```
then choose on witch render we want to plot by using the argument `at`
```python
 plt.show(object, at=0) # plot on first renderer
``` 
If we want to add a slider it takes the last renderer to be used. However the `pos` keyword is relative to the full window (not the selected renderer) but takes origin in the selected renderer. (which is confusing)
```python
plt = vedo.Plotter(N=2,...)
plt.show(object,at =1)
plt.addSlider2d(slider_func, ..., pos= [(0.05,0.04), (0.15,0.04)])
```
will show the slider on the second renderer. The issue is if we set a pos value greater than 0.5, it is leaking from the window (ex: `pos= (0.4,0.04), (0.6,0.04)` will display only one half of the slider)

In order to choose on which renderer you will show the slider one can use a empty `show`, for example :
```python
# some previous code
plt.show(object,at=1)
plt.show(at=0)
plt.addSlider2d(slider_func, ..., pos= [(0.05,0.04), (0.15,0.04)])
```
will show the slider on the first one. 
I am not sure on how bad practice it is, and maybe @marcomusy will have a better way to do so.

**Comment by marcomusy** [comment, marked as answer]
Hi Anton, sorry for the late reply, I need to look at how things can be improved on this because indeed it's confusing..
the `plt.show(at=0)` is perfectly fine, maybe one can replace it with `plt.renderer = plt.renderers[0]` to make the first renderer the "current" renderer (I haven't tested it!).

**Reply by antonfrancois** [reply]
Yes it works, and executes less code, so I guess it is better ? Thanks anyways, thats refreshing to have your support !

**Comment by marcomusy** [comment]
BTW this could be a solution with multiple renderers:
```python
from vedo import *

settings.immediateRendering = False  # faster for multi-renderers

# (0,0) is the bottom-left corner of the window, (1,1) the top-right
# the order in the list defines the priority when overlapping
custom_shape = [
    dict(bottomleft=(0.0, 0.0), topright=(0.5, 1), bg="wheat", bg2="w"),  # ren0
    dict(bottomleft=(0.5, 0.0), topright=(1, 1), bg="blue3", bg2="lb"),  # ren1
    dict(bottomleft=(0.2, 0.05), topright=(0.8, 0.1), bg="white"),  # ren2
]
plt = Plotter(shape=custom_shape, size=(1600, 900))

s0 = ParametricShape(0)
s1 = ParametricShape(1)

plt.show(s0, "renderer0", at=0)
plt.show(s1, "renderer1", at=1)


def slider1(widget, event):
    value = sliderRep.GetValue()
    s0.rotateY(value)
    s1.rotateY(-value)


plt.renderer = plt.renderers[2]  # make it the current renderer
slider = plt.addSlider2D(slider1, -5, 5, value=0, pos=([0.05, 0.02], [0.55, 0.02]))
sliderRep = slider.GetRepresentation()
vscale = 20
sliderRep.SetSliderLength(0.003 * vscale)  # make it thicker
sliderRep.SetSliderWidth(0.025 * vscale)
sliderRep.SetEndCapLength(0.001 * vscale)
sliderRep.SetEndCapWidth(0.025 * vscale)
sliderRep.SetTubeWidth(0.0075 * vscale)

interactive().close()
```

![Screenshot from 2021-11-16 13-09-44](https://user-images.githubusercontent.com/32848391/141983112-01505790-b9de-47ca-8389-3647417f9934.png)

which makes me realize that I could again improve on clarity on various things :)
so thanks for drawing my attention to it.
Cheers
M.

**Reply by antonfrancois** [reply]
Yes it is indeed much closer to the end result I was expected ! However as it really shrink the sliders. As I need to show four of them, is there a faster way than applying

>
> ``` 
> vscale = 20
>sliderRep.SetSliderLength(0.003 * vscale)  # make it thicker
>sliderRep.SetSliderWidth(0.025 * vscale)
>sliderRep.SetEndCapLength(0.001 * vscale)
>sliderRep.SetEndCapWidth(0.025 * vscale)
>sliderRep.SetTubeWidth(0.0075 * vscale)
>```
to each slider (by passing a dictionary of smth) ?  Or maybe adding a slider_size multiplier would be of great use =)
Cheers

**Reply by marcomusy** [reply]
yes - I was planning to pass a dictionary... hope i can do it in the next release! if you have many sliders you may write a function like `mystyle(sliderRep)`

**Reply by antonfrancois** [reply]
It is what I ended up doing.  By the way I didn't found a Set function for title position, is it tweak-able ?

**Reply by marcomusy** [reply]
this: https://vtk.org/doc/nightly/html/classvtkSliderRepresentation2D.html
gives all what is possible to tweak from the base class

**Comment by HeavenlyBerserker** [comment]
Hi, thanks for these responses! They were very helpful. I can't get the value on the slider to display. It is too small and there doesn't seem to be a parameter that changes this. Any ideas?

**Reply by marcomusy** [reply]
would this be a solution:

```py
from vedo import Plotter, Mesh, dataurl, get_color_name

mesh = Mesh(dataurl+"magnolia.vtk")

plt = Plotter()
plt += [mesh]

def sliderfunc(widget, event):
    mesh.color(widget.value)
    cname = get_color_name(mesh.color())
    widget.title = cname

plt.add_slider(
    sliderfunc,
    xmin=-9,
    xmax=9,
    value=0,
    pos="top-right",
    title="change color",
    title_height=0.05,
    show_value=False,
    font="Calco",
)
plt.show()
```
<img width="324" alt="Screenshot 2023-05-24 at 15 38 38" src="https://github.com/marcomusy/vedo/assets/32848391/d547ef8d-990d-4a8f-ae6a-670461fd1fe8">
